linter: --lsp mode does not apply plugins re-added via extends (CLI works, --print-config correct, LSP returns empty

### What version of Oxlint are you using?

1.67.0 (also reproduced with 1.66.0)

### What command did you run?

`oxlint --lsp` (launched by neovim's nvim-lspconfig oxlint server, but reproduces with any LSP client). CLI comparison: oxlint app/path/to/file.ts from the child config directory.

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

Workspace root : .oxlintrc.json:

```jsonc
{
  "$schema": "./node_modules/oxlint/configuration_schema.json",
  "plugins": ["import", "promise", "node"],
  "categories": {
    "correctness": "error",
    "suspicious": "warn",
    "perf": "warn"
  },
  "rules": {
    "typescript/consistent-type-imports": ["warn", { "prefer": "type-imports" }]
  }
}
```

Child : apps/api/.oxlintrc.json:

```jsonc
{
  "$schema": "../../node_modules/oxlint/configuration_schema.json",
  "extends": ["../../.oxlintrc.json"],
  "ignorePatterns": ["database/schema.ts", ".adonisjs"],
  "rules": {
    "unicorn/filename-case": ["error", { "case": "snakeCase" }]
  }
}

```

### What happened?

When the root config declares an explicit plugins array, the default-on plugins (typescript, unicorn, oxc) are dropped from that config : --print-config from the root confirms this. The child config has no plugins field, so via extends the defaults are merged back in, and --print-config from apps/api/ correctly shows all six plugins.

### CLI behaves consistently with --print-config:

```bash
# From apps/api/  (extends merges defaults back in)
$ oxlint app/auth/guard/firebase.ts
app/auth/guard/firebase.ts:5:1: warning typescript(consistent-type-imports): ...
app/auth/guard/firebase.ts:6:1: warning typescript(consistent-type-imports): ...
app/auth/guard/firebase.ts:8:1: warning typescript(consistent-type-imports): ...
app/auth/guard/firebase.ts:11:30: error typescript(no-wrapper-object-types): ...
```

### LSP does not:

Launching oxlint --lsp with cwd = apps/api and sending textDocument/diagnostic for the exact same file returns:

{ "items": [], "kind": "full" }

Verified via /proc/<pid>/cwd that the LSP process really is running in apps/api, and via --print-config from that same cwd that the resolved config includes the typescript/unicorn/oxc plugins. So the LSP is reading the right config on disk, but is not applying the plugins that were re-added through extends. it behaves as if it were using the parent config's literal plugins array.

### Expected Behavior

oxlint --lsp should produce the same diagnostics as the CLI when given the same cwd, the same binary, and the same config tree, in particular, plugins re-added via extends (by virtue of the child config omitting plugins) should be active.

### Actual Behavior

LSP returns zero diagnostics; CLI returns four. --print-config agrees with the CLI.

### Workaround

Add the default plugins explicitly to the root config so neither path relies on the extends merge:

"plugins": ["import", "promise", "node", "typescript", "unicorn", "oxc"]

### Related

Possibly the same root cause as the (closed) #21030, which also reports --print-config disagreeing with lint-time behavior under nested/extended configs. That issue was closed without a stated fix and the underlying mismatch still reproduces on 1.67.0.

You can find a minimal reproduction [here](https://github.com/kerwanp/oxc-linter-lsp-repro). (disclaimer: done with AI)

I spotted two inconsistencies, one might be expected behavior.

This might be expected behavior but this feel incoherent. When running `oxlint` from root, nested configs are ignored, the same files results with different diagnostics whether cwd is `./` or `./apps/api`.

This behavior also occurs with `--lsp` even tho the working directory is `./apps/api`.

## Reproduction

###  Running `oxlint` from root

No issues found.

```bash
❯ ./node_modules/.bin/oxlint
Found 0 warnings and 0 errors.
Finished in 20ms on 3 files using 32 threads.
```

### Running `oxlint` from `apps/api`

2 issues found.

```bash
❯ ../../node_modules/.bin/oxlint --format=stylish

oxc-repro/apps/api/src/example.ts
  1:1   warning  All imports in the declaration are only used as types. Use `import type`.  typescript(consistent-type-imports)
  7:29  error  Do not use wrapper object types.  typescript(no-wrapper-object-types)
```

### Running `oxlint --lsp` in `apps/api`

No issues found.

```
❯ node lsp-repro.mjs
LSP returned 0 diagnostic(s):
```

--- a/apps/oxlint/fixtures/lsp/issue_22758/.oxlintrc.json
+++ b/apps/oxlint/fixtures/lsp/issue_22758/.oxlintrc.json
@@ -1,0 +1,6 @@
+{
+  "plugins": ["import", "promise", "node"],
+  "categories": {
+    "correctness": "warn"
+  }
+}

--- a/apps/oxlint/fixtures/lsp/issue_22758/apps/api/.oxlintrc.json
+++ b/apps/oxlint/fixtures/lsp/issue_22758/apps/api/.oxlintrc.json
@@ -1,0 +1,3 @@
+{
+  "extends": ["../../.oxlintrc.json"]
+}

--- a/apps/oxlint/fixtures/lsp/issue_22758/apps/api/app/auth/guard/firebase.ts
+++ b/apps/oxlint/fixtures/lsp/issue_22758/apps/api/app/auth/guard/firebase.ts
@@ -1,0 +1,1 @@
+let _value: String;

--- a/apps/oxlint/src/config_loader.rs
+++ b/apps/oxlint/src/config_loader.rs
@@ -257,6 +257,10 @@ pub struct LoadedConfigs {
     pub nested_ignore_patterns: Vec<(Vec<String>, PathBuf)>,
 }
 
+pub fn materialize_default_plugins(config: &mut Oxlintrc) {
+    config.plugins.get_or_insert_with(Default::default);
+}
+
 pub struct ConfigLoader<'a> {
     external_linter: Option<&'a ExternalLinter>,
     external_plugin_store: &'a mut ExternalPluginStore,

--- a/apps/oxlint/src/lint.rs
+++ b/apps/oxlint/src/lint.rs
@@ -24,7 +24,9 @@ use crate::{
     cli::{
         CliRunResult, DebugOption, LintCommand, MiscOptions, ReportUnusedDirectives, WarningOptions,
     },
-    config_loader::{CliConfigLoadError, ConfigLoadError, ConfigLoader},
+    config_loader::{
+        CliConfigLoadError, ConfigLoadError, ConfigLoader, materialize_default_plugins,
+    },
     output_formatter::{LintCommandInfo, OutputFormatter},
     walk::Walk,
 };
@@ -282,11 +284,10 @@ impl CliRunner {
             }
         };
 
-        {
-            let mut plugins = root_config.plugins.unwrap_or_default();
-            enable_plugins.apply_overrides(&mut plugins);
-            root_config.plugins = Some(plugins);
-        }
+        materialize_default_plugins(&mut root_config);
+        let mut plugins = root_config.plugins.unwrap_or_default();
+        enable_plugins.apply_overrides(&mut plugins);
+        root_config.plugins = Some(plugins);
 
         let base_ignore_patterns = root_config.ignore_patterns.clone();
 

--- a/apps/oxlint/src/lsp/server_linter.rs
+++ b/apps/oxlint/src/lsp/server_linter.rs
@@ -31,6 +31,7 @@ use oxc_language_server::{
 use crate::{
     config_loader::{
         ConfigLoader, build_nested_configs, config_file_names, discover_configs_in_tree,
+        materialize_default_plugins,
     },
     lsp::{
         code_actions::{
@@ -105,13 +106,14 @@ impl ServerLinterBuilder {
         #[cfg(feature = "napi")]
         let loader = loader.with_js_config_loader(self.js_config_loader.as_ref());
 
-        let oxlintrc = match loader.load_root_config(&root_path, config_path.as_ref()) {
+        let mut oxlintrc = match loader.load_root_config(&root_path, config_path.as_ref()) {
             Ok(config) => config,
             Err(e) => {
                 warn!("Failed to load config: {e}");
                 Oxlintrc::default()
             }
         };
+        materialize_default_plugins(&mut oxlintrc);
 
         let mut nested_ignore_patterns = Vec::new();
         let mut extended_paths = FxHashSet::default();
@@ -1518,5 +1520,11 @@ mod test {
             }),
         );
         tester.test_and_snapshot_single_file("test.ts");
+    }
+
+    #[test]
+    fn test_issue_22758() {
+        let tester = Tester::new("fixtures/lsp/issue_22758/apps/api", json!({}));
+        tester.test_and_snapshot_single_file("app/auth/guard/firebase.ts");
     }
 }

